exonum/src/api/mod.rs

```diff
@@ -316,89 +205,142 @@ pub trait ExtendApiBackend {
-                    (format!("services/{}", name), builder)
-                }),
-        );
+        let context = ApiContext::with_blockchain(&blockchain);
+        endpoints.extend(dispatcher.api_endpoints(&context));
 
-        trace!("Create actix-web worker with api: {:#?}", inner);
+        trace!("Create actix-web worker with api: {:#?}", endpoints);
 
         match access {
             ApiAccess::Public => backend.extend(
-                inner
+                endpoints
                     .iter()
                     .map(|(name, builder)| (name.as_ref(), &builder.public_scope)),
             ),
             ApiAccess::Private => backend.extend(
-                inner
+                endpoints
                     .iter()
                     .map(|(name, builder)| (name.as_ref(), &builder.private_scope)),
             ),
         }
     }
 
-    /// Adds API factory with the given prefix to the aggregator.
-    pub fn insert<S: Into<String>>(&mut self, prefix: S, builder: ServiceApiBuilder) {
-        self.inner.insert(prefix.into(), builder);
+    /// Add API factory with the given prefix to the aggregator.
+    pub fn insert<S: Into<String>>(&mut self, prefix: S, builder: ApiBuilder) {
+        self.endpoints.insert(prefix.into(), builder);
     }
 
-    /// Refreshes shared node state.
+    /// Refresh shared node state.
     fn refresh(&self) {
         self.node_state.update_dispatcher_state(&self.blockchain);
     }
 
-    fn explorer_api(
-        blockchain: &Blockchain,
-        shared_node_state: SharedNodeState,
-    ) -> ServiceApiBuilder {
-        let mut builder = ServiceApiBuilder::new();
-        let service_api_state = ServiceApiState::new(blockchain.clone());
-        ExplorerApi::wire(builder.public_scope(), service_api_state, shared_node_state);
+    fn explorer_api(context: ApiContext, shared_node_state: SharedNodeState) -> ApiBuilder {
+        let mut builder = ApiBuilder::new();
+
+        ExplorerApi::new(context).wire(builder.public_scope(), shared_node_state);
         builder
     }
 
-    fn system_api(shared_api_state: SharedNodeState) -> ServiceApiBuilder {
-        // Waits until dispatcher will be unlocked to get fresh info.
-        let mut builder = ServiceApiBuilder::new();
-        self::node::private::SystemApi::new(NodeInfo::new(), shared_api_state.clone())
+    fn system_api(context: ApiContext, shared_api_state: SharedNodeState) -> ApiBuilder {
+        let mut builder = ApiBuilder::new();
+
+        let sender = context.sender().clone();
+        self::node::private::SystemApi::new(sender, NodeInfo::new(), shared_api_state.clone())
             .wire(builder.private_scope());
-        self::node::public::SystemApi::new(shared_api_state).wire(builder.public_scope());
+        self::node::public::SystemApi::new(context, shared_api_state).wire(builder.public_scope());
         builder
     }
 }
+
+/// Provide the current state of the blockchain to the API handlers.
```

Comment:
NItpick: since this is a struct and not a function, I'd suggest "This entity provides".